### PR TITLE
Add plugin-defined settings through `ConfigurationContribution`s

### PR DESCRIPTION
**Pull request by BadPrograms**
# Add plugin preferences support

## Overview

This PR introduces **plugin preferences**, which work similarly to the existing napari settings system. (docs are in napari/npe2#502)

Plugin preferences are stored separately for each plugin and can be accessed programmatically through a new `get_plugin_settings()` API, providing a consistent way for plugins to define, persist, and react to user configuration.

Each plugin has its own settings file, stored in the same directory as the napari settings. Plugin preferences are also displayed in the Preferences dialog below the main napari preferences.

Plugin preferences

---

## Configuration example

Plugins can define configuration entries similarly to napari settings:

```yaml
configurations:
  reader:
    title: Reading with something
    properties:
      lazy:
        type: boolean
        default: false
        title: Load lazily
  writer:
    title: Writing
    properties:
      lazy:
        type: boolean
        default: false
        title: write lazily
      other_Attribute:
        type: boolean
        default: false
        title: something
```

Configuration titles and property names are converted into Python-accessible attributes. Therefore, identifiers cannot contain spaces, dashes, or other invalid characters; use underscores instead (e.g., plugin_parameter, not plugin-parameter).

---

## Accessing plugin preferences

All plugin preferences can be accessed through:

```python
from napari.settings import get_plugin_settings

get_plugin_settings()
```

or for a specific plugin:

```python
get_plugin_settings("test-plugin")
```

This returns a `PluginPreferences` object, which is based on `EventedModel`. This allows plugin preferences to expose events when fields are modified.

---

## Listening to preference changes

Preference change events can be accessed through the `events` attribute:

```python
settings = get_plugin_settings("test-plugin")

settings.reader.events.connect(
    callback_func
)
###OR###
settings.events.["reader"].connect(callback_func)
```

Plugins that do not need reactive behavior can simply read the stored values directly using:
```python
settings = get_plugin_settings("test-plugin")

plugin_variable = settings.reader.lazy
```

---

## Implementation changes

This PR introduces:

* `PluginSettings` as a management class 

* A new `get_plugin_settings()` API, similar to `get_settings()`

* Methods for generating preference widgets dynamically

* Plugin-specific settings storage and loading

* Event support for plugin preference fields

---
## Limitations

Dynamically created Pydantic models do not support autocomplete yet.

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
This looks really cool and useful! I know people have wanted plugin settings managed in the GUI for a long time.
General question, which shows my ignorance of this, because it might just be an existing npe2 thing:
- will a plugin with settings in manifest work in older napari that doesn't have this GUI? so will it just silently fall back?

For followup: I think it could be nicer to make the UI work like the Shortcuts? So there would be one position in the left-hand list e.g. Plugin Settings and then the plugins would be selectable on the right side?
Ideally we'd have a keyboard shortcut to go direct to Plugin Settings like we do for Shortcuts.

**Comment by BadPrograms** [comment]
Hello Peter,

Yes, having configurations in the manifest has been supported since npe2, but napari itself did not make use of them. As a result, older versions of napari simply won’t display the new preference widgets.

During the hackathon, the general consensus was to use this approach instead of the shortcuts-style stack widget, although I don’t remember the exact reasoning behind that decision.

Soon, I will add the required tests to finish the PR.

**Comment by psobolewskiPhD** [comment]
Sounds good.

> During the hackathon, the general consensus was to use this approach instead of the shortcuts-style stack widget, although I don’t remember the exact reasoning behind that decision.

Thanks for the context. I think we can consider advantages and disadvantages in followup. I think the keybind to Plugins Settings like for Shortcuts is a big advantage of that approach, but maybe there is something obvious I'm missing.

**Comment by brisvag** [comment]
> During the hackathon, the general consensus was to use this approach instead of the shortcuts-style stack widget, although I don’t remember the exact reasoning behind that decision.

I think I was (and am) with @psobolewskiPhD on this, but definitely let's get this in and iterate :+1:

**Comment by BadPrograms** [comment]
I also created a parent class for NapariSettings and PluginPreferences called GeneralSettings.

**Comment by TimMonko** [comment]
In [`aa3eadf` (this PR)](https://github.com/napari/napari/pull/9308/commits/aa3eadfe57f5d9085968c142b24929bab2cec9c3), Giannis and I debugged the `min_req` fail from the earlier tests. It was a result of lack of name normalization for the title _and_ eager property checks by pydantic <2.10. We opted for name normlization (like `field_name`) as it does seem useful to pass a valid python name, but did want to discuss our process whereby this _could_ be solved with a bump to pydantic>=2.10 because there it is stored as a dict key and not eagerly accessed -- and as such it will only break if `inspect.signature({name})` is called. I cna't find that we sanitize elsewhere though, but I guess its because our settings are not declarative like the manifest?

I'm going to do a full review now

**Comment by TimMonko** [comment]
`min_req` is passing. Docs fail is unrelated 😁 
I've made napari/npe2#501 to demonstrate the manifest validation for the config contribution to prevent name clashes. If we did that, then we can bump to npe2 0.8.4 (we need to make a release anyways) and it's a nice little clean up change: https://github.com/TimMonko/napari/commit/24be72d1ec8485984a71adedb8693056f9141a04 <- this can be a follow-up, certainly.

**Comment by TimMonko** [comment]
Ok, I'll need to do another logic review pass, but I stumbled across mostly some formatting/dead code things that needed to get changed so I made BadPrograms/napari#2 -- it reverts some formatting that were hiding real changes.
The only logic change I needed was in the snake case identifier for the path of longer python names.

**Comment by TimMonko** [comment]
So I think things are working, I just needed to work through some thought processes and ultimately landed on napari/npe2#502. If this docs doesn't sufficiently describe how everything works -- or it reveals the way of something working that we don't like, then we need to fix things. I haven't otherwise reviewed this GUI implementation, yesterday's review was focused on cleanup. That said, I have it working nicely from my own plugin so far and yeah it took a bit to understand the quirks which is why I focused on docs today.

**Comment by TimMonko** [comment]
@BadPrograms, could you update the PR description now that its near its in its final form and review napari/npe2#502?
I've tested its implementation thoroughly in ndev-kit/ndevio#71 and it seems to be doing everything well. Tests took a bit to figure out how they would be used because of how npe2's pytest fixture works (it blocks manifest discovery, which is good), but I stuck that all in the documentation -- so you can read there for more. 
So I think I'm approving expect for my few questions that popped up in napari/npe2#502, namely: do we want to `snake_case` the category title without giving it a name like for the properties? I think what we have is ok. Otherwise, see if there are any smells in those docs

**Comment by TimMonko** [comment]
CircleCI docs fail are unrelated, and a requests issue

File: src/napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py [hunk 1]
```diff
@@ -784,7 +784,10 @@ def populate_from_schema(
         layout.setFieldGrowthPolicy(QtWidgets.QFormLayout.FieldGrowthPolicy(1))
         for name, sub_schema in schema['properties'].items():
             label, widget = self._prepare_widget(name, sub_schema, widget_builder, ui_schema)
-            if len(schema['properties']) == 1:
+            # render a label unless the property explicitly has an empty title
+            # (intentional divergence from upstream qt_json_builder, which
+            # suppressed the label for any single-property schema)
+            if sub_schema.get('title') == '':
```
Comment (TimMonko) [inline src/napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py:790]: it remains a question to me if we want to change vendored code. I'm not sure of our code practices -- @DragaDoncila ?
If we don't do this, then we effectively have to rebuild the class on our own to override it, which also feels a bit silly

File: src/napari/settings/_plugin_config_generator.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    # camelCase -> snake_case (e.g. someSetting -> some_Setting)
+    name = re.sub(r'(?<!^)(?=[A-Z])', '_', name)
+    # any run of non-alphanumeric characters is a separator
+    name = re.sub(r'[^0-9a-zA-Z]+', '_', name)
+    name = re.sub(r'_+', '_', name).strip('_').lower()
+    if not name:
+        name = 'settings'
+    if name[0].isdigit():
+        name = f'_{name}'
+    return name
```
Comment (TimMonko) [inline src/napari/settings/_plugin_config_generator.py]: this can be competely removed because we now know that config and property keys are valid Python identifiers with the npe2 changes. The key should (and will) be used verbatim

File: src/napari/settings/_plugin_config_generator.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+}
+# npe2 coerces configuration property types to JSON Schema type names
+# (see ``npe2.manifest.contributions._json_schema``), so only those four
+# need mapping
+_TYPE_MAP: dict[str, type] = {
+    'boolean': bool,
+    'integer': int,
+    'number': float,
+    'string': str,
+}
```
Comment (TimMonko) [inline src/napari/settings/_plugin_config_generator.py]: This can/should be completely removed because npe2's ConfigurationJsonSchema.python_type (via _python_equivalent) does this mapping for us

File: src/napari/settings/_napari_settings.py
Comment (TimMonko) [inline src/napari/settings/_napari_settings.py:1]: I think, but I'm not sure, that we actually don't want to abstract `GeneralSettings` and should thus revert this file (and others effected by this change, like `_migrations.py`).
This is because we can only maintain CURRENT_SCHEMA_VERSION in one place, because I don't think we've fully hooked up the migration machinery for plugins (nor should we?) And I think we don't want the plugin yamls to hold napari's version?

File: src/napari/settings/_plugin_config_generator.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+
+def _build_single_config_model(
```
Comment (TimMonko) [inline src/napari/settings/_plugin_config_generator.py:85]: this method simplifies greatly with removal of _TYPE_MAP and _snake_identifier()

File: src/napari/settings/_plugin_config_generator.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+) -> dict[str, type[PluginPreferences]]:
+    if not plugin_manager:
+        from npe2 import PluginManager
+
+        pm = PluginManager.instance()
+        pm.discover()
+    else:
+        pm = plugin_manager
+    plugins = sorted(
+        pm.iter_manifests(),
```
Comment (TimMonko) [inline src/napari/settings/_plugin_config_generator.py]:
```suggestion
        pm.iter_manifests(disabled=False),
```
Skip disabled plugins to match other harvesting behavior

**Review by TimMonko — COMMENTED** [review]
submitting a review here for a moment -- about to chat to Draga quick
the only other thing is caching the original state and restoring on "cancel" of the settings like in napari's own settings
I will probably make a branch and make a PR, but you'd be welcome to accept changes as you see fit

File: src/napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py (hunk 1 above)
Comment (TimMonko) [reply src/napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py:790]: we discussed this as ok

File: src/napari/settings/_plugin_config_generator.py (hunk 4 above)
Comment (TimMonko) [reply src/napari/settings/_plugin_config_generator.py:85]: done

**Comment by TimMonko** [comment]
I just pushed a small, sort of unrelated change (https://github.com/napari/napari/pull/9308/commits/1d7d882337631bd0e6208379d6435f6efffceaad) to remove the stretch (I feel like stretch is unintuitive in Qt) from the pages list so that it expands. It's only relevant now that these entries are contributable and cause it to need to be scrolled quickly, like seen in OP description. Here is what it looks like now:
image

File: src/napari/settings/_plugin_config_generator.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        extra = {
+            k: v
+            for k, v in data.items()
+            if VALUE_TRANSLATOR.get(k, k) not in _FIELD_KWARGS
+        }
+        if extra:
+            field_kwargs['json_schema_extra'] = {
+                **field_kwargs.get('json_schema_extra', {}),
+                **extra,
+            }
```
Comment (TimMonko) [inline src/napari/settings/_plugin_config_generator.py:131]: the logic change I introduced here is due to a pydantic deprecation warning that we filter out in napari's pyproject.toml. Fields used to have `extra`, but now its put in as `json_schema_extra`. I've updated this because I hit the edge case when testing enum on ndevio's manifest

**Review by TimMonko — APPROVED** [review]
@BadPrograms, I think this is ready to go, great work! Draga and I worked on test fixtures and finalizing things on the npe2 side.

**Comment by TimMonko** [comment]
Worked to expand test coverage and ran into an issue in the re-implementation of the `NAPARI_CONFIG` path / settings path logic (via a new `_PL_CONFIG_DIR`). I've now settled it on sharing the same logic as napari's general settings. https://github.com/napari/napari/pull/9308/commits/1e28bab0fdd75bfbb7cebe4944222c54f47b3500
Other than that I introduced two dead branches now that we have type gaurantees from npe2 and added a few tests.
If tests pass and I didn't actually miss coverage than I think indeed we are at ready-to-merge :)

**Comment by TimMonko** [comment]
We've hit [100% coverage](https://app.codecov.io/gh/napari/napari/pull/9308?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=napari) (the comment is out of date it seems)

Currently, the preferences reset button behavior we hooked into resets ALL of napari settings and plugin settings. This is ok as a first pass IMO. But we should consider adding a per-page reset. This would benefit both napari (e.g. you could just reset experimental) and plugins (you reset only each plugin).

**Comment by TimMonko** [comment]
@czaki besides adding a docstring, I made some changes it would be good to have your eyes on:
1. I centralized the `mock_pm` fixture to conftest https://github.com/napari/napari/pull/9308/commits/507ff0cfa9695ac8cb93e364a1a8b0ee3b87b748
2. rely on the autouse `plugin_settings_` fixture https://github.com/napari/napari/pull/9308/commits/b47004c6d6200449c948b6db8816b40e3ff925f5
3. added enum validation with pydantic (since it otherwise doesn't validate enum values set programmatically) https://github.com/napari/napari/pull/9308/commits/ac2e86cc49bac3d848755e02969fbbc8a7fda855

**Comment by Czaki** [comment]
For me the PR is ready to merge. Feel free to do this.

**Comment by TimMonko** [comment]
Green is go! Thank you @BadPrograms and @DragaDoncila for all your incredible work on this! So excited to see this fully realized!
